File: ios/Packages/PadelCore/Sources/PadelCore/Sync/SessionPayload.swift
import Foundation

// DTOs del contrato con la app de liga (`docs/api-contract.md`).
//
// Están separados del modelo de dominio a propósito: el modelo puede evolucionar sin
// romper el contrato, y el contrato puede versionarse sin contaminar el dominio.

public struct SessionPayload: Codable, Equatable, Sendable {
    public let sessionId: String
    public let schemaVersion: Int
    public let source: SourcePayload
    public let startedAt: String
    public let endedAt: String
    public let durationSeconds: Int64
    public let player: PlayerPayload?
    public let matchRef: MatchRefPayload?
    public let shots: ShotsPayload
    public let health: HealthPayload?
    public let score: ScorePayload?
    public let level: LevelPayload?
}

/// Nivel técnico estimado, en la escala de pádel de 1 a 7.
///
/// Es **derivado**: la liga puede recalcularlo de los golpeos si algún día quiere usar su
/// propia fórmula. Viaja en el payload para que no tenga que hacerlo.
public struct LevelPayload: Codable, Equatable, Sendable {
    public let overall: Float
    public let byShotType: [String: Float]
    /// 0 a 1. En pádel la regularidad es tanto del nivel como la potencia.
    public let consistency: Float
    public let repertoire: Float
    public let gradedShots: Int
    /// false si hubo pocos golpeos: el número está, pero no hay que fiarse de él.
    public let reliable: Bool
}

public struct ScorePayload: Codable, Equatable, Sendable {
    public let rules: ScoreRulesPayload
    public let sets: [SetScorePayload]
    /// "us" | "them", o ausente si el partido no llegó a terminarse.
    public let winner: String?
    public let completed: Bool
}

public struct ScoreRulesPayload: Codable, Equatable, Sendable {
    /// "advantage" | "goldenPoint" | "starPoint".
    public let deuceFormat: String
    public let setsToWin: Int
}

public struct SetScorePayload: Codable, Equatable, Sendable {
    public let us: Int
    public let them: Int
}

public struct SourcePayload: Codable, Equatable, Sendable {
    public let platform: String
    public let device: String
    public let appVersion: String
}

public struct PlayerPayload: Codable, Equatable, Sendable {
    public let hand: String
    public let watchWrist: String
}

public struct MatchRefPayload: Codable, Equatable, Sendable {
    public let matchId: String
    public let leagueId: String?
}

public struct ShotsPayload: Codable, Equatable, Sendable {
    public let total: Int
    public let byType: [String: Int]
    public let intensity: IntensityPayload?
    public let events: [ShotEventPayload]
}

public struct IntensityPayload: Codable, Equatable, Sendable {
    public let meanRacketSpeedKmh: Float
    public let maxRacketSpeedKmh: Float
    public let meanImpactG: Float
    public let maxImpactG: Float
}

public struct ShotEventPayload: Codable, Equatable, Sendable {
    public let offsetMs: Int64
    public let type: String
    public let racketSpeedKmh: Float
    public let impactG: Float
    public let confidence: Float
}

public struct HealthPayload: Codable, Equatable, Sendable {
    public let heartRate: HeartRatePayload?
    public let activeEnergyKcal: Float?
    public let totalEnergyKcal: Float?
    public let steps: Int?
    public let distanceMeters: Float?
    public let zonesSeconds: [String: Int]?
}

public struct HeartRatePayload: Codable, Equatable, Sendable {
    public let meanBpm: Int
    public let maxBpm: Int
    public let restingBpm: Int?
}

public struct SessionRefResponse: Codable, Sendable {
    public let id: String
    public let sessionId: String?
    public let createdAt: String?
    public let matchRef: MatchRefPayload?
}

public struct APIErrorResponse: Codable, Sendable {
    public let error: APIErrorBody?
}

public struct APIErrorBody: Codable, Sendable {
    public let code: String?
    public let message: String?
}

public struct MatchesResponse: Codable, Sendable {
    public let matches: [MatchSummary]
}

public struct MatchSummary: Codable, Equatable, Sendable, Identifiable {
    public let matchId: String
    public let leagueId: String?
    public let scheduledAt: String?
    public let opponents: String?
    public let venue: String?

    public var id: String { matchId }
}

extension PadelSession {
    /// Convierte la sesión al payload del contrato.
    ///
    /// - Parameter shareHealth: consentimiento del usuario. Si es false el bloque
    ///   `health` no se incluye en absoluto (no se manda vacío: se omite).
    /// - Parameter includeEvents: si es false solo se suben los agregados. Se usa para
    ///   reintentar una sesión que el servidor rechazó por tamaño.
    public func toPayload(shareHealth: Bool, includeEvents: Bool = true) -> SessionPayload {
        SessionPayload(
            sessionId: sessionId,
            schemaVersion: schemaVersion,
            source: SourcePayload(
                platform: source.platform.wireName,
                device: source.device,
                appVersion: source.appVersion
            ),
            startedAt: isoUTC(startedAtEpochMs),
            endedAt: isoUTC(endedAtEpochMs),
            durationSeconds: durationSeconds,
            player: PlayerPayload(
                hand: profile.hand.wireName,
                watchWrist: profile.watchWrist.wireName
            ),
            matchRef: matchRef.map { MatchRefPayload(matchId: $0.matchId, leagueId: $0.leagueId) },
            shots: ShotsPayload(
                total: totalShots,
                byType: shotsByType.reduce(into: [String: Int]()) { result, entry in
                    result[entry.key.wireName] = entry.value
                },
                intensity: shots.isEmpty ? nil : IntensityPayload(
                    meanRacketSpeedKmh: round1(intensity.meanRacketSpeedKmh),
                    maxRacketSpeedKmh: round1(intensity.maxRacketSpeedKmh),
                    meanImpactG: round1(intensity.meanImpactG),
                    maxImpactG: round1(intensity.maxImpactG)
                ),
                events: includeEvents
                    ? shots.map { shot in
                        ShotEventPayload(
                            offsetMs: shot.offsetMs,
                            type: shot.type.wireName,
                            racketSpeedKmh: round1(shot.racketSpeedKmh),
                            impactG: round1(shot.impactG),
                            confidence: round2(shot.confidence)
                        )
                    }
                    : []
            ),
            health: shareHealth ? health.toPayload() : nil,
            score: score.map { $0.toPayload() },
            level: level.toPayload()
        )
    }
}

extension SessionLevel {
    /// Sin golpeos puntuables no hay nivel que mandar: se omite en vez de mandar un 1 falso.
    func toPayload() -> LevelPayload? {
        guard gradedShots > 0 else { return nil }
        return LevelPayload(
            overall: round1(overall),
            byShotType: Dictionary(
                uniqueKeysWithValues: byShotType.map { ($0.key.wireName, round1($0.value)) }
            ),
            consistency: round2(consistency),
            repertoire: round2(repertoire),
            gradedShots: gradedShots,
            reliable: reliable
        )
    }
}

extension MatchScore {
    func toPayload() -> ScorePayload {
        ScorePayload(
            rules: ScoreRulesPayload(
                deuceFormat: rules.deuceFormat.wireName,
                setsToWin: rules.setsToWin
            ),
            sets: allSets.map { SetScorePayload(us: $0.us, them: $0.them) },
            winner: winner?.wireName,
            completed: isFinished
        )
    }
}

extension HealthMetrics {
    func toPayload() -> HealthPayload? {
        guard !isEmpty else { return nil }
        return HealthPayload(
            heartRate: heartRate.map {
                HeartRatePayload(meanBpm: $0.meanBpm, maxBpm: $0.maxBpm, restingBpm: $0.restingBpm)
            },
            activeEnergyKcal: activeEnergyKcal.map(round1),
            totalEnergyKcal: totalEnergyKcal.map(round1),
            steps: steps,
            distanceMeters: distanceMeters.map(round1),
            zonesSeconds: zones.secondsPerZone.isEmpty ? nil : zones.secondsPerZone
        )
    }
}

/// ISO-8601 UTC truncado a segundos: `2026-07-25T18:04:12Z`.
///
/// Sin pasar por `Int`: en los relojes arm64_32 `Int` es de 32 bits, y aunque los
/// segundos de época caben hoy, dejan de caber en 2038. Con épocas, siempre `Int64`.
func isoUTC(_ epochMs: Int64) -> String {
    let seconds = (Double(epochMs) / 1000).rounded(.down)
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime]
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    return formatter.string(from: Date(timeIntervalSince1970: seconds))
}

func round1(_ value: Float) -> Float { (value * 10).rounded() / 10 }

func round2(_ value: Float) -> Float { (value * 100).rounded() / 100 }

File: android/core/src/main/kotlin/com/risingpadel/core/sync/SessionPayload.kt
package com.risingpadel.core.sync

import com.risingpadel.core.analytics.SessionAnalytics
import com.risingpadel.core.level.SessionLevel
import com.risingpadel.core.model.HealthMetrics
import com.risingpadel.core.model.PadelSession
import com.risingpadel.core.score.MatchScore
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant
import java.time.format.DateTimeFormatter
import kotlin.math.round

/**
 * DTOs del contrato con la app de liga (`docs/api-contract.md`).
 *
 * Están separados del modelo de dominio a propósito: el modelo puede evolucionar sin
 * romper el contrato, y el contrato puede versionarse sin contaminar el dominio.
 */
@Serializable
data class SessionPayload(
    val sessionId: String,
    val schemaVersion: Int,
    val source: SourcePayload,
    val startedAt: String,
    val endedAt: String,
    val durationSeconds: Long,
    val player: PlayerPayload? = null,
    val matchRef: MatchRefPayload? = null,
    val shots: ShotsPayload,
    val health: HealthPayload? = null,
    val score: ScorePayload? = null,
    val level: LevelPayload? = null,
    val analytics: AnalyticsPayload? = null,
)

/**
 * Series agregadas para las gráficas de la liga. Opcional y **aditivo**: no sube la
 * versión del esquema, y un receptor viejo lo ignora sin romperse.
 *
 * Son decenas de números y no los eventos por golpeo: caben en un deep link y no
 * exponen nada que los agregados (nivel, recuentos) no expongan ya.
 */
@Serializable
data class AnalyticsPayload(
    val frequency: FrequencyPayload? = null,
    val levelProgression: LevelProgressionPayload? = null,
    /**
     * Nivel medio del jugador sobre su historial en el dispositivo emisor, si lo conoce.
     * Es la línea "calidad media" de las gráficas de la liga.
     */
    val playerAverageLevel: Float? = null,
)

@Serializable
data class FrequencyPayload(
    val intervalMinutes: Int,
    /** Golpeos por intervalo, desde el minuto 0. Los intervalos vacíos van como 0. */
    val counts: List<Int>,
)

@Serializable
data class LevelProgressionPayload(
    val points: List<LevelProgressionPointPayload>,
)

@Serializable
data class LevelProgressionPointPayload(
    /** Minuto de sesión en el que se evalúa el punto (fin de la ventana). */
    val minute: Float,
    val level: Float,
)

/**
 * Nivel técnico estimado, en la escala de pádel de 1 a 7.
 *
 * Es **derivado**: la liga puede recalcularlo de los golpeos si algún día quiere usar su
 * propia fórmula. Viaja en el payload para que no tenga que hacerlo.
 */
@Serializable
data class LevelPayload(
    val overall: Float,
    val byShotType: Map<String, Float>,
    /** 0 a 1. En pádel la regularidad es tanto del nivel como la potencia. */
    val consistency: Float,
    val repertoire: Float,
    val gradedShots: Int,
    /** false si hubo pocos golpeos: el número está, pero no hay que fiarse de él. */
    val reliable: Boolean,
)

@Serializable
data class SourcePayload(
    val platform: String,
    val device: String,
    val appVersion: String,
)

@Serializable
data class PlayerPayload(
    val hand: String,
    val watchWrist: String,
)

@Serializable
data class MatchRefPayload(
    val matchId: String,
    val leagueId: String? = null,
)

@Serializable
data class ShotsPayload(
    val total: Int,
    val byType: Map<String, Int>,
    val intensity: IntensityPayload? = null,
    val events: List<ShotEventPayload> = emptyList(),
)

@Serializable
data class IntensityPayload(
    val meanRacketSpeedKmh: Float,
    val maxRacketSpeedKmh: Float,
    val meanImpactG: Float,
    val maxImpactG: Float,
)

@Serializable
data class ShotEventPayload(
    val offsetMs: Long,
    val type: String,
    val racketSpeedKmh: Float,
    val impactG: Float,
    val confidence: Float,
)

@Serializable
data class HealthPayload(
    val heartRate: HeartRatePayload? = null,
    val activeEnergyKcal: Float? = null,
    val totalEnergyKcal: Float? = null,
    val steps: Int? = null,
    val distanceMeters: Float? = null,
    val zonesSeconds: Map<String, Int>? = null,
)

@Serializable
data class ScorePayload(
    val rules: ScoreRulesPayload,
    val sets: List<SetScorePayload>,
    /** "us" | "them", o ausente si el partido no llegó a terminarse. */
    val winner: String? = null,
    val completed: Boolean,
)

@Serializable
data class ScoreRulesPayload(
    /** "advantage" | "goldenPoint" | "starPoint". */
    val deuceFormat: String,
    val setsToWin: Int,
)

@Serializable
data class SetScorePayload(val us: Int, val them: Int)

@Serializable
data class HeartRatePayload(
    val meanBpm: Int,
    val maxBpm: Int,
    val restingBpm: Int? = null,
)

/**
 * Estado en vivo de un partido, para que otros lo sigan mientras se juega.
 *
 * Es **estado completo, no eventos**: cada actualización sustituye del todo a la
 * anterior, así que perder una no rompe nada — la siguiente trae la verdad entera. Por
 * eso se publica con PUT y sin cola de reintentos: reenviar un marcador viejo sería
 * peor que no enviar nada.
 */
@Serializable
data class LiveScorePayload(
    val sessionId: String,
    val updatedAt: String,
    /** true en la última publicación: el partido acabó y el espectador deja de refrescar. */
    val completed: Boolean,
    val score: ScorePayload? = null,
    val shotCount: Int = 0,
    val heartRateBpm: Int? = null,
    val elapsedSeconds: Long = 0,
)

@Serializable
data class SessionRefResponse(
    val id: String,
    val sessionId: String? = null,
    val createdAt: String? = null,
    val matchRef: MatchRefPayload? = null,
)

@Serializable
data class ApiErrorResponse(val error: ApiErrorBody? = null)

@Serializable
data class ApiErrorBody(val code: String? = null, val message: String? = null)

@Serializable
data class MatchesResponse(val matches: List<MatchSummary> = emptyList())

@Serializable
data class MatchSummary(
    val matchId: String,
    val leagueId: String? = null,
    val scheduledAt: String? = null,
    val opponents: String? = null,
    val venue: String? = null,
)

/**
 * Convierte una sesión al payload del contrato.
 *
 * @param shareHealth consentimiento del usuario. Si es false el bloque `health` no se
 *   incluye en absoluto (no se manda vacío: se omite).
 * @param includeEvents si es false solo se suben los agregados. Se usa para reintentar
 *   una sesión que el servidor rechazó por tamaño.
 * @param playerAverageLevel media del historial del jugador, si quien llama la conoce.
 *   La sesión no sabe de historial: por eso entra como parámetro.
 */
fun PadelSession.toPayload(
    shareHealth: Boolean,
    includeEvents: Boolean = true,
    playerAverageLevel: Float? = null,
): SessionPayload = SessionPayload(
    sessionId = sessionId,
    schemaVersion = schemaVersion,
    source = SourcePayload(
        platform = source.platform.wireName,
        device = source.device,
        appVersion = source.appVersion,
    ),
    startedAt = isoUtc(startedAtEpochMs),
    endedAt = isoUtc(endedAtEpochMs),
    durationSeconds = durationSeconds,
    player = PlayerPayload(hand = profile.hand.wireName, watchWrist = profile.watchWrist.wireName),
    matchRef = matchRef?.let { MatchRefPayload(it.matchId, it.leagueId) },
    shots = ShotsPayload(
        total = totalShots,
        byType = shotsByType.entries.associate { (type, count) -> type.wireName to count },
        intensity = if (shots.isEmpty()) null else intensity.let {
            IntensityPayload(
                meanRacketSpeedKmh = round1(it.meanRacketSpeedKmh),
                maxRacketSpeedKmh = round1(it.maxRacketSpeedKmh),
                meanImpactG = round1(it.meanImpactG),
                maxImpactG = round1(it.maxImpactG),
            )
        },
        events = if (!includeEvents) emptyList() else shots.map { shot ->
            ShotEventPayload(
                offsetMs = shot.offsetMs,
                type = shot.type.wireName,
                racketSpeedKmh = round1(shot.racketSpeedKmh),
                impactG = round1(shot.impactG),
                confidence = round2(shot.confidence),
            )
        },
    ),
    health = if (shareHealth) health.toPayloadOrNull() else null,
    score = score?.toPayload(),
    level = level.toPayloadOrNull(),
    analytics = analyticsPayloadOrNull(playerAverageLevel),
)

/** Sin golpeos no hay series que mandar: se omite el bloque entero. */
private fun PadelSession.analyticsPayloadOrNull(playerAverageLevel: Float?): AnalyticsPayload? {
    if (shots.isEmpty()) return null
    val analytics = SessionAnalytics()
    val durationMs = durationSeconds * 1000

    val buckets = analytics.shotFrequency(shots, durationMs, SessionAnalytics.INTERVAL_10_MIN_MS)
    val points = analytics.levelProgression(shots, durationMs)

    return AnalyticsPayload(
        frequency = buckets.takeIf { it.isNotEmpty() }?.let { list ->
            FrequencyPayload(intervalMinutes = 10, counts = list.map { it.count })
        },
        levelProgression = points.takeIf { it.isNotEmpty() }?.let { list ->
            LevelProgressionPayload(
                points = list.map {
                    LevelProgressionPointPayload(
                        minute = round1(it.offsetMs / 60_000f),
                        level = round1(it.level),
                    )
                }
            )
        },
        playerAverageLevel = playerAverageLevel?.let { round2(it) },
    )
}

/** Sin golpeos puntuables no hay nivel que mandar: se omite en vez de mandar un 1 falso. */
private fun SessionLevel.toPayloadOrNull(): LevelPayload? {
    if (gradedShots == 0) return null
    return LevelPayload(
        overall = round1(overall),
        byShotType = byShotType.entries.associate { (type, value) -> type.wireName to round1(value) },
        consistency = round2(consistency),
        repertoire = round2(repertoire),
        gradedShots = gradedShots,
        reliable = reliable,
    )
}

private fun MatchScore.toPayload() = ScorePayload(
    rules = ScoreRulesPayload(
        deuceFormat = rules.deuceFormat.wireName,
        setsToWin = rules.setsToWin,
    ),
    sets = allSets.map { SetScorePayload(us = it.us, them = it.them) },
    winner = winner?.wireName,
    completed = isFinished,
)

private fun HealthMetrics.toPayloadOrNull(): HealthPayload? {
    if (isEmpty) return null
    return HealthPayload(
        heartRate = heartRate?.let { HeartRatePayload(it.meanBpm, it.maxBpm, it.restingBpm) },
        activeEnergyKcal = activeEnergyKcal?.let { round1(it) },
        totalEnergyKcal = totalEnergyKcal?.let { round1(it) },
        steps = steps,
        distanceMeters = distanceMeters?.let { round1(it) },
        zonesSeconds = zones.secondsPerZone.takeIf { it.isNotEmpty() },
    )
}

/** ISO-8601 UTC truncado a segundos: `2026-07-25T18:04:12Z`. */
internal fun isoUtc(epochMs: Long): String =
    DateTimeFormatter.ISO_INSTANT.format(Instant.ofEpochSecond(Math.floorDiv(epochMs, 1000L)))

private fun round1(value: Float): Float = round(value * 10f) / 10f

private fun round2(value: Float): Float = round(value * 100f) / 100f
